build-tools/scripts/RunCmake.proj

```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Prepare" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="Cmake"
+      DependsOnTargets="_ValidateCmake;_RunCmake"
+  />
+  <Target Name="_ValidateCmake">
+    <Error
+        Condition=" '$(CmakePath)' == '' "
+        Text="Set the `%24(CmakePath)` property."
+    />
+    <Error
+        Condition=" '$(CmakeGenerator)' == '' "
+        Text="Set the `%24(CmakeGenerator)` property."
+    />
+    <Error
+        Condition=" '$(CmakeSourceDir)' == '' "
+        Text="Set the `%24(CmakeSourceDir)` property."
+    />
+    <Error
+        Condition=" '$(CmakeBuildDir)' == '' "
+        Text="Set the `%24(CmakeBuildDir)` property."
+    />
+  </Target>
+
+  <Target Name="_RunCmake">
+    <PropertyGroup>
+      <_Prepare>$(PrepareNativeToolchain)</_Prepare>
+      <_Prepare Condition=" '$(_Prepare)' != '' And !$(_Prepare.Trim().EndsWith('&amp;&amp;')) ">$(_Prepare) &amp;&amp;</_Prepare>
+      <_SourceDir>$(CmakeSourceDir.Replace('%5c', '/'))</_SourceDir>
+      <_BuildDir>$(CmakeBuildDir.Replace('%5c', '/'))</_BuildDir>
+      <_ExtraArgs>$(CmakeExtraArgs.Replace('%5c', '/'))</_ExtraArgs>
```

Comment:
Since this is already working, this is fine probably fine. But was the main issue here the `<Exec/>` task?

Maybe it would work to make a small `<UnixExec/>` task that subclasses `<Exec/>` and replaces all the `\` with `/` in `Command`? Then `base.Execute()` would succeed?

Reply:
@jonathanpeppers : the problem isn't "unix", the problem is double quotes. :-)

"General guidelines", at least on Unix, is to quote anything which could possibly have a space…such as path names!  The problem is when:

1. The "path name" is a directory name, and
2. The directory name *ends with* `\`.

The result is that you *escape the quote*!

```
$(_Prepare) $(CmakePath) $(CmakeGenerator) -S &quot;$(_SourceDir)&quot;
```

Here, `$(_SourceDir)` will be e.g. `$(MSBuildThisFileDirectory)`, which *aways* ends with `\`.  If we didn't "fix" `$(_SourceDir)`, we'd have:

```
make -S "C:\some\directory\" …
```

The attempt to "sanely" quote the directory name results in an invalid command line.

Does this mean that `s,\,/,g` is "best"? No; we could instead do:

```xml
<_SourceDir>$(CmakeSourceDir.TrimEnd('%5c'))</_SourceDir>
```

But!  If a value can be "part of" the Cmake files, e.g. Cmake variables, those values *must* use `/`, not `\`, because Cmake otherwise attempts to treat `\` as an escape character.  Thus, `$(CmakeExtraArgs)` *must* replace `\` with `/`, as it contains Cmake property overrides that contain paths.

In neither case does a `<UnixExec/>` class help, unless a "command global" `s,\,/,g` is desired.  Note however that we *don't* "fixup" `$(PrepareNativeToolchain)` or `$(CmakePath)`.